### PR TITLE
Rework common.mk

### PR DESCRIPTION
**Pull request by pascal-brand-st-dev**
The code added by commits d6536da64deb ("Commonlize xtest compilation
among all platforms") and 070d95595e1d1 ("common build: optee
components") introduces race conditions that break parallel build (make
-jX). This is due to manually launching a make sub-process to run the
-common targets, and can be fixed by just declaring a dependency
instead. Specific flags are appended to common flags, rather than
passed as arguments to a new "make" sub-process.

So, basically:

 foo-common:
 	$(MAKE) -C foo_dir [A=a B=B]

 foo:
 	$(MAKE) foo-common [C=c]

is turned into:

 FOO_COMMON_FLAGS ?= A=a B=b
 foo-common:
 	$(MAKE) -C foo_dir $(FOO_COMMON_FLAGS)

 FOO_COMMON_FLAGS += C=c
 foo: foo-common

At the same time this allows foo: to override flags, for instance
set A=aa in the above example (prior to this, A=a would have had a
higher priority).

Signed-off-by: Jerome Forissier <jerome.forissier@linaro.org>

### CODE REVIEW

File: qemu.mk [hunk 1]
```diff
@@ -195,7 +191,8 @@ endef
 
 .PHONY: run
 # This target enforces updating root fs etc
-run: | bios-qemu run-only
+run: bios-qemu
```
Comment (jforissier) [inline qemu.mk]: You need qemu and soc-term, too. Or just use: `run: all`.

**Comment by jforissier** [comment]
@pascal-brand-st-dev thanks for creating the PR.

With the comment I made re. `qemu.mk`, and for the files that were not in my original patch (i.e., `fvp.mk`, `hikey.mk` and `mediatek.mk`):

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

You should perhaps add your `Signed-off-by`, since you wrote parts of this patch?

**Comment by pascal-brand-st-dev** [comment]
Updated

**Comment by pascal-brand-st-dev** [comment]
* hikey / mediatek / qemu are compiling fine with `make -j8 all`
* Still some issues on fvp with parallel build.

Should I merge anyway?

**Comment by jforissier** [comment]
We'd better fix the remaining issue(s) I think. I have a QEMU environment, is there a way to switch to FVP without cloning everything again? Then I can take a look if you wish.

**Comment by jbech-linaro** [comment]
@jforissier I was switching back and forth when creating the manifest and usually it goes without too much trouble. If you have files that has been modified (kernel, qemu I think), then I recommend that you run `git checkout -f && git clean -xdf` before switching. When you are about to switch, just simply run
`repo init ....` with the fvp xml file instead, and then just run `repo sync` again.

By the way, consider setting up a reference for future use as described in our instructions. It save a huge amount of time.

**Comment by jforissier** [comment]
@jbech-linaro thanks for the clarification. I'll setup a reference and let it clone during the week end.

As I was reading README.md I thought maybe this reference could be used only for one platform (i.e., have a fixed set of git repositories and just switch branches or commits). Now I understand that subprojects and references/branches/commits are all defined in the manifest files, so it makes sense one can switch as desired.

**Comment by pascal-brand-st-dev** [comment]
@jforissier I'll work on this then.

**Comment by pascal-brand-st-dev** [comment]
`make -j8 run` is working fine on fvp. So now parallel compilation can be used on all platforms.

**Comment by jforissier** [comment]
Apart from the style issues which you may or may not change, as you prefer:
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`
